src/webgpu/api/validation/capability_checks/limits/limit_utils.ts

```diff
@@ -0,0 +1,292 @@
+    device.pushErrorScope('out-of-memory');
+    device.pushErrorScope('validation');
+
+    const result = fn({ ...deviceAndLimits, testValueName, testValue, shouldError });
+    if (result instanceof Promise) {
+      await result;
+    }
+
+    const validationError = await device.popErrorScope();
+    const outOfMemoryError = await device.popErrorScope();
+    const internalError = await device.popErrorScope();
+
+    this.expect(!validationError, validationError?.message || '');
+    this.expect(!outOfMemoryError, outOfMemoryError?.message || '');
+    this.expect(!internalError, internalError?.message || '');
+
+    device.destroy();
+    this._device = undefined;
+  }
+
+  /**
+   * Calls a function that expects a GPU error if shouldError is true
+   */
+  async expectGPUError<R>(
+    filter: GPUErrorFilter,
+    fn: () => R,
+    shouldError: boolean = true,
+    msg = ''
+  ): Promise<R> {
+    const { device } = this;
+
+    device.pushErrorScope(filter);
+    const returnValue = fn();
+    if (returnValue instanceof Promise) {
+      await returnValue;
+    }
+
+    const error = await device.popErrorScope();
+    this.expect(
+      !!error === shouldError,
+      `${error?.message || 'no error when one was expected'}: ${msg}`
+    );
+
+    return returnValue;
+  }
+
+  /**
+   * Calls a function that expects a validation error if shouldError is true
+   */
+  async expectValidationError<R>(fn: () => R, shouldError: boolean = true, msg = ''): Promise<R> {
+    return this.expectGPUError('validation', fn, shouldError, msg);
+  }
+
+  /**
+   * Calls a function that expects to not generate a validation error
+   */
+  async expectNoValidationError<R>(fn: () => R, msg = ''): Promise<R> {
+    return this.expectGPUError('validation', fn, false, msg);
+  }
+
+  /**
+   * Calls a function that might expect a validation error.
+   * if shouldError is true then expect a validation error,
+   * if shouldError is false then ignore out-of-memory errors.
+   */
+  async testForValidationErrorWithPossibleOutOfMemoryError<R>(
+    fn: () => R,
+    shouldError: boolean = true,
+    msg = ''
+  ): Promise<R> {
+    const { device } = this;
+
+    if (!shouldError) {
+      device.pushErrorScope('out-of-memory');
+      const result = fn();
+      await device.popErrorScope();
+      return result;
+    }
+
+    // Validation should fail before out-of-memory so there is no need to check
+    // for out-of-memory here.
+    device.pushErrorScope('validation');
+    const returnValue = fn();
+    const validationError = await device.popErrorScope();
+
+    this.expect(
+      !!validationError,
+      `${validationError?.message || 'no error when one was expected'}: ${msg}`
+    );
+
+    return returnValue;
+  }
+}
+
+/**
+ * Makes a new LimitTest class so that the tests have access to `limit`
+ */
+function makeLimitTestFixture(limit: GPUSupportedLimit): typeof LimitTestsImpl {
+  class LimitTests extends LimitTestsImpl {
+    limit = limit;
```

Comment:
Possible to write this out like a mixin to avoid playing this trick with overriding the parent field `limit`?

Reply:
I'm not sure what you mean by a mixin here

I want the base class to have a `limit` field, not the derived class, In fact I don't want a derived class. But, as `makeTestGroup` takes a constuctor instead of a factory function I have no way to pass the limit field to the place `LimitTests` will actually be constructed.

Reply:
It's not important, maybe I'll play to see if there's a possible improvement but most likely not.

Reply:
I looked into making TestGroup the a FixtureClassFactory and then refactored makeTestGroup take either a FixtureClass or a FictureClassFactory and if it wasn't a factory then generate an anonymous one.

But, after a couple of hours I couldn't get it to be happy so I punted for now. I'm sure it's possible.